src/react/react-decor-function.ts

```diff
@@ -0,0 +1,107 @@
+import React = require('react');
+import {
+    Attributes,
+    ReactNode,
+    ReactHTML,
+    ReactSVG,
+    SFC,
+    ComponentClass,
+    ClassType,
+    ClassicComponent,
+    cloneElement,
+    ComponentState,
+    ClassicComponentClass
+} from 'react';
+import { after, before, FunctionWrapper, middleware, MiddlewareHook } from '../function-decor';
+import { THList, THListToTuple } from "typelevel-ts";
+import { cloneDeep } from 'lodash';
+
+
+export type ElementHook = <P = object>(componentProps: any, args: ElementArgs<P>) => ElementArgs<P>;
+export type RootNodeHook = <P = object>(componentProps: any, rootNodeProps: Attributes & Partial<P>) => Attributes & Partial<P>;
+
+export type ElementArgs<P extends {}> = {
+    type: any,
+    elementProps: Attributes & Partial<P>,
+    children: Array<ReactNode>
+}
+export type CreateElementArgsTuple<P extends {}> = [any, Attributes & Partial<P>, ReactNode];
+
+type ReactCreateElement = typeof React.createElement;
+
+const originalCreateElement = React.createElement;
+
+export interface DecorReactHooks {
+    nodes?: Array<ElementHook>
+    root?: Array<ElementHook>
+}
+
+export function decorReact(hooks: DecorReactHooks): Function {
+    const mw = (next: Function, args: [object]) => {
+        let renderResult, createArgsMap: Map<object, ElementArgs<any>> = new Map();
+
+        try {
+            const componentProps = args[0]; // [0] for props in a functional react component
+            let wrappedCreateElement = makeCustomCreateElement(hooks, componentProps, createArgsMap);
+            React.createElement = wrappedCreateElement;
+
+            renderResult = next(args);
+
+            if (hooks.root && hooks.root.length > 0) {
+                let rootElementArgs = createArgsMap.get(renderResult);
+
+                rootElementArgs = hooks.root.reduce(<P extends {}>(elementArgs: ElementArgs<P>, hook: ElementHook): ElementArgs<P> => hook(componentProps, elementArgs), rootElementArgs);
+
+                if (rootElementArgs) {
+                    renderResult = cloneElement(renderResult, rootElementArgs.elementProps);
+                }
+            }
+        } finally {
+            React.createElement = originalCreateElement as ReactCreateElement;
+            createArgsMap.clear();
+        }
+
+        return renderResult;
+    }
+    return middleware(mw);
+}
+
+// create a custom react create element function that applies the given hooks
+function makeCustomCreateElement<P extends {}>(hooks: DecorReactHooks, componentProps: object, createArgsMap: Map<object, ElementArgs<P>>): typeof originalCreateElement {
+    let wrappedCreateElement = originalCreateElement;
+
+    const executeHook = (res: ElementArgs<any>, hook: Function): ElementArgs<any> => {
```

Comment:
I'd give this function a name that hints better as to its designed role: it's a reducer function for hooks array.

Also, this function can be defined globally once, outside `makeCustomCreateElement`. I think it can be also used for the root element reduce in line 53